File: src/main/Python/nowcoder/S49_StrToInt.py
class StrToInt:
    def strToInt(self,str):
        if len(str) == 0 or not str:
            return 0
        start = 0
        symbol = 1
        res = 0
        if str[0] == '+':
            start = 1
        elif str[0] == '-':
            start = 1
            symbol = -1
        for i in range(start, len(str)):
            if str[i] >= '0' and str[i] <= '9':
                res = res * 10 + ((ord(str[i]) - ord('0')))
            else:
                return 0
        if res > 2147483647:
            return 0
        return symbol * res
if __name__ == '__main__':
    test = StrToInt()
    print(test.strToInt("-2147483649"))

File: src/main/java/nowcoder/S49_StrToInt.java
package nowcoder;

import scala.Int;

/**
 * @author LiuZhiguo
 * @date 2020/2/11 11:47
 */
public class S49_StrToInt {
    public int strToInt(String str){
        if (str == null || str.length() == 0)
            return 0;
        char[] chars = str.trim().toCharArray();
        int start = 0;
        int symbol = 1;
        int res = 0;
        int temp = 0;
        if (chars[start] == '+')
            start = 1;
        else if (chars[start] == '-'){
            start = 1;
            symbol = -1;
        }
        for (int i=start;i<chars.length;i++){
            if (chars[i] < '0' || chars[i] > '9')
                return 0;
            //当前各位数字
            int digit = chars[i] - '0';
            temp = res * 10 + digit;
            //判断正数是否溢出
            if (symbol == 1 && res * 10 > Integer.MAX_VALUE - digit)
                return 0;
            //判断负数是否溢出
            if (symbol == -1 && res * 10 < Integer.MIN_VALUE -digit)
                return 0;
            res = temp;
        }
        return res * symbol;
    }
    public static void main(String[] args){
        S49_StrToInt s49 = new S49_StrToInt();
        System.out.println(s49.strToInt("-2147483649"));
    }
}
